Upgrade to electron 12 and Node 14

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

This PR upgrades our electron version to the current latest 12.X and the current latest node 14.X.

I tried upgrading to electron 13.X but ran into trouble with the `@electron/remote` library. I think we should probably try and remove that before upgrading as it is considered harmful to security.

Notes:

- I have removed `spectron` as our integration testing library as it doesn't work with subframes in electron@12.X. 
- I have rewrote our integration tests using `playwright`. This did require changing some of the tests because playwright doesn't have some of the indepth electron APIs. I believe that this leads to better tests because the interactions that the tests can use are closer to what the users can do.

Nokel81 committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

Upgrade to electron 12 and Node 14

Signed-off-by: Sebastian Malton <sebastian@malton.name>

# This is the commit message #2:

Fix tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on package.json:
```diff
@@ -365,7 +367,7 @@
     "sass-loader": "^8.0.2",
     "sharp": "^0.26.1",
-    "spectron": "11.0.0",
+    "spectron": "14",
```
Seems that we don't need spectron anymore, probably should be removed from here?

jakolehm commented on extensions/pod-menu/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lens-pod-menu",
-  "version": "0.0.1",
+  "version": "5.1.4-alpha.1628102440259.1628102553246",
```
These are probably added by accident.

aleksfront commented on src/renderer/components/+apps-helm-charts/helm-charts.tsx:
```diff
@@ -86,17 +86,19 @@ export class HelmCharts extends Component<Props> {
               placeholder: "Search Helm Charts...",
             },
+            ...headerPlaceholders
           })}
```
Helm Charts page was designed without header title and expanded search box by purpose. This changes breaks initial layout.

Before the changes (looks like expected):
![helm charts before](https://user-images.githubusercontent.com/9607060/129560416-6be15964-5d7b-470a-bf5c-cbf34b509696.png)

Current view (broken):
![helm charts after](https://user-images.githubusercontent.com/9607060/129560437-a57e8e21-0454-4192-8f45-96d62d4eeddd.png)

We can just discard changes in this file.

Nokel81 replied:
Okay, I will have to revisit the tests then because I needed that `Helm Charts` title.

Nokel81 committed:
Upgrade to electron 12 and Node 14

- Compute cluster ID for tests
- Switch to temporary appData path while testing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

nevalla commented on integration/helpers/utils.ts:
```diff
@@ -69,71 +34,45 @@ export function describeIf(condition: boolean) {
-  name: string;
-  url: string;
-};
+  const minikubeFrame = await window.waitForSelector("#cluster-frame-484e864bad9b84ce5d6b4fff704cc0e4");
```
where does this id come from?

Nokel81 replied:
From line 205 in `src/main/catalog-sources/kubeconfig-sync.ts`:

```ts
const clusterId = createHash("md5").update(`${filePath}:${contextName}`).digest("hex");
```

I guess I should probably make it so that the the ID is computed on demand for each OS

nevalla commented on src/renderer/components/+apps-releases/releases.tsx:
```diff
@@ -138,7 +138,7 @@ export class HelmReleases extends Component<Props> {
             ...headerPlaceholders,
           })}
-          renderHeaderTitle="Releases"
+          renderHeaderTitle="Helm Releases"
```
This change is not related to Electron version bump(?)

Nokel81 replied:
No and I will revert

nevalla commented on src/renderer/utils/createStorage.ts:
```diff
@@ -47,15 +46,11 @@ export function createStorage<T>(key: string, defaultValue: T) {
-  }
-
-  function init() {
+  if (!storage.initialized && !process.argv.includes("--integration-testing")) {
```
ideally this `!process.argv.includes("--integration-testing")` should be in helper method that is called here

Nokel81 replied:
Sounds good

nevalla commented on src/renderer/ipc/index.tsx:
```diff
@@ -117,7 +117,7 @@ function HotbarTooManyItemsHandler(): void {
 }
 
-export function registerIpcHandlers() {
+export function registerIpcListeners() {
```
Not sure if this change is needed either, but probably makes sense better now.

nevalla reviewed (dismissed):
Generally looking good and could not see any errors while testing this. For some reason integration tests are failing. I triggered re-run, but if those will fail again, there must be some real issue. 

We should probably have helper method to resolve is app running on integration tests to reduce code repetition.

jakolehm commented on src/main/window-manager.ts:
```diff
@@ -75,6 +75,12 @@ export class WindowManager extends Singleton {
       const { width, height, x, y } = this.windowState;
+      const additionalArguments = [];
+
+      if (isIntegrationTesting) {
```
Why do we need this?

jakolehm commented on src/renderer/components/+apps-helm-charts/helm-charts.tsx:
```diff
@@ -86,6 +86,7 @@ export class HelmCharts extends Component<Props> {
             [columnId.name]: chart => chart.getName(),
             [columnId.repo]: chart => chart.getRepository(),
           }}
+          renderHeaderTitle="Helm Charts"
```
Unrelated change?

jakolehm commented on src/renderer/utils/createStorage.ts:
```diff
@@ -22,16 +22,16 @@
 
 const storage = observable({
-  initialized: false,
+  initialized: isIntegrationTesting,
```
This feels odd, we should not change logic while running integration tests.

Nokel81 replied:
Fair enough, the changing of userData path is probably better.

jakolehm commented on integration/helpers/utils.ts:
```diff
@@ -69,71 +37,69 @@ export function describeIf(condition: boolean) {
-      CICD: "true"
+  for (let i = 0; i < 10; i += 1) {
+    try {
+      return await electron.launch({
```
This is missing temporary path for user data. Now it's pretty dangerous (and probably flaky) to run integration tests because it will read/modify users real environment. Most likely `--integration-testing` flag can be removed after we fix this.

Nokel81 replied:
Not quite, but we don't need to pass it to the renderer if we change the user data path. I initially introduced this flag so as to not block the app from quitting (which we do under normal circumstances).

I guess a proper solution would be to send a SIGTERM to the process, I might look into that.

jakolehm reviewed (dismissed):
We need to use temporary path for user data during tests.

jakolehm commented on build/download_kubectl.ts:
```diff
@@ -117,7 +117,7 @@ class KubectlDownloader {
 
 const downloadVersion = packageInfo.config.bundledKubectlVersion;
-const baseDir = path.join(process.env.INIT_CWD, "binaries", "client");
+const baseDir = path.join(__dirname, "..", "binaries", "client");
```
What's the reason behind this change?

Nokel81 replied:
Because on windows `INIT_CWD` is not defined.

jakolehm replied:
For future readers: `INIT_CWD` is not defined anymore on node14.

jakolehm commented on integration/__tests__/app.tests.ts:
```diff
@@ -25,63 +25,55 @@
-// FIXME (!): improve / simplify all css-selectors + use [data-test-id="some-id"] (already used in some tests below)
-describe("Lens integration tests", () => {
-  let app: Application;
+    try {
```
Try/finally should be done using before/afterEach to avoid repetition in tests.

jakolehm commented on src/common/ipc/ipc.ts:
```diff
@@ -46,6 +48,7 @@ function getSubFrames(): ClusterFrameInfo[] {
 }
 
 export function broadcastMessage(channel: string, ...args: any[]) {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
```
Why this is needed? 🤔

Nokel81 replied:
I don't know, I think a relic of a revert from master.

jakolehm commented on src/common/utils/getPath.ts:
```diff
@@ -0,0 +1,38 @@
+    return app.getPath(name);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
```
Eslint-disable seems odd also here, is it required?

jakolehm commented on src/renderer/utils/createStorage.ts:
```diff
@@ -68,32 +63,35 @@ export function createAppStorage<T>(key: string, defaultValue: T, clusterId?: st
-    });
+    reaction(
+      () => toJS(storage.data),
+      async (state: Record<string, any> = {}) => {
```
Was this change necessary? Usually code blocks are refactored to smaller logical chunks, this goes opposite way.

Nokel81 committed:
some cleanup

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Switch to using test hooks

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/main/menu.ts:
```diff
@@ -80,10 +81,11 @@ export function buildMenu(windowManager: WindowManager) {
       { type: "separator" },
       {
         label: "Preferences",
+        id: "Preferences",
```
Why do some get an `id` and not others (like Extensions)? For testing purposes?

Nokel81 replied:
I will be removing these once I get CI to pass as I no longer need them. Hopefully in the future we will be able to interact with the menu but it doesn't seem to work yet.

Commit:
Fix build, upgrade to 12.0.17

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

jim-docker commented on extensions/kube-object-event-status/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "kube-object-event-status",
-  "version": "0.0.1",
+  "version": "5.2.0-beta.2.1629813675276",
```
still need to revert this?

Nokel81 replied:
Yup

jim-docker commented on integration/__tests__/app-preferences.tests.ts:
```diff
@@ -0,0 +1,74 @@
+    ({ window, cleanup } = await utils.start());
+    await utils.clickWelcomeButton(window);
+    await window.keyboard.press("Meta+,");
+  }, 10*60*1000);
```
10 minutes seems rather long?

Nokel81 replied:
Fair however each statement has a 30s timeout by default. And I would much prefer it if `playwright` times out then jest because the error messages are much better.

jim-docker commented on src/common/utils/getPath.ts:
```diff
@@ -0,0 +1,37 @@
+/**
+ * calls getPath either on app or on the remote's app
+ *
+ * @deprecated Use a different method for accessing the getPath function
```
Adding a new function that is immediately deprecated? What different method should be used?

Nokel81 replied:
Given that these shouldn't be changed once they are set. The better solution would be to initialize them for each frame once. But I wanted to leave that sort of thing to follow up PRs

jakolehm replied:
Deprecated new function sounds strange but let's fix this in follow-up PR.

Nokel81 committed:
Simplify cleanup

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
revert to using app.close() and catch remove tmpDir errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on integration/helpers/utils.ts:
```diff
@@ -69,71 +40,77 @@ export function describeIf(condition: boolean) {
+    args: ["--integration-testing"], // this argument turns off the blocking of quit
+    executablePath: AppPaths[process.platform],
+    bypassCSP: true,
+    env: { CICD },
```
what does this do, just sets one env var, an empty path?

Nokel81 replied:
It shouldn't be empty. It should be a part to a new temporary directory. This allows each integration test to start with a clean slate in terms of local storage (so dock tabs and the sidebar open status for instance).

jakolehm committed:
fix helm repos integration tests

Signed-off-by: Jari Kolehmainen <jari.kolehmainen@gmail.com>

Nokel81 committed:
Remove wait for list helm repos

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
revert extension files

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove unneeded menu IDs

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "build:mac": "yarn run compile && electron-builder --mac --dir",
     "build:win": "yarn run compile && electron-builder --win --dir",
-    "integration": "jest --runInBand integration",
+    "integration": "jest --runInBand --detectOpenHandles --forceExit integration",
```
I hope we can remove `--forceExit` in follow-up PRs.

Nokel81 replied:
Yes I completely agree, I am going to try and upgrade to the newer version of Jest in a follow-up PR.

jakolehm commented on src/main/index.ts:
```diff
@@ -209,7 +211,7 @@ app.on("ready", async () => {
   const windowManager = WindowManager.createInstance();
 
-  cleanup.push(
+  onQuitCleanup.push(
```
Note: these fixes are also in a separate PR.

Nokel81 replied:
Shall we merge that other one first and then rebase this one?

jakolehm commented on src/renderer/utils/createStorage.ts:
```diff
@@ -68,32 +63,35 @@ export function createAppStorage<T>(key: string, defaultValue: T, clusterId?: st
-    });
+    reaction(
+      () => toJS(storage.data),
+      async state => {
```
My previous comment has somehow got lost. This change does not really make sense to me. It does not alter the behaviour but just does the opposite of usual refactoring.. it bundles more logic into one (usually we refactor code blocks to smaller chunks).

Nokel81 replied:
I will revert this and put it into a separate PR. I disagree that it is bundling more logic because I am removing function declarations which are only ever used once.

jakolehm replied:
Just noticed that those functions were inline functions. Now the refactoring makes more sense.. anyway better to do outside of this PR because it is an unrelated change.

Nokel81 committed:
revert createStorage changes

Signed-off-by: Sebastian Malton <sebastian@malton.name>
